### ISSUE TITLE
Sending async requests to FastAPI endpoint

### ISSUE THREAD

**Issue by assman** [tiangolo/fastapi#3309]
I have a POST path operation function defined in my app which generates OTPs for users. Here's the pseudo code - 

```
@app.post("/auth/generate-token", status_code=status.HTTP_201_CREATED)
def create_otp(user_email: models.UserEmailModel, response: Response):
    # Firstly, I check the database if there is an existing non-expired OTP that belongs to the user
    non_expired_otp = check_for_non_expired_otp(user_email.email_address)
    if non_expired_otp:
        return "OTP is still valid/not expired"

    # Otherwise, generate an OTP and email it to the user.
    new_otp = generate_otp_for_user(user_email.email_address)
    return "success"
```

It works fine in general but the problem is, when I use an HTTP client to send multiple(like 4-5) async POST requests to the endpoint, the endpoint generates 4 different OTPs. Seems like the function is running asynchronously, but it's causing unintended results since I only wish to generate one OTP for a user at a time instead of multiple. Would greatly appreciate if someone could point me in the right direction.

**Comment by KiraPC** [comment]
It seems that the endpoint is correct.

Are you sure that the bug is not on the client code?

**Comment by assman** [comment]
> It seems that the endpoint is correct.
> 
> 
> 
> Are you sure that the bug is not on the client code?

The client code is just calling this endpoint 4 times asynchronously. It's the endpoint that's running the path operation function asynchronously that's causing trouble for me. I wrote some simple debug print statements in the endpoint to confirm that.

Moreover, when the client calls the endpoint 4 times synchronously, this behavior is not seen because it's sending the requests one after the other.

**Comment by panla** [comment]
hello.

the doc is `https://fastapi.tiangolo.com/zh/async/`

```text
you can mix def and async def in your path operation functions as much as you need and define each one using the best option for you. FastAPI will do the right thing with them.
Anyway, in any of the cases above, FastAPI will still work asynchronously and be extremely fast.
```

**Comment by assman** [comment]
> https://fastapi.tiangolo.com/zh/async/

Yes I have read through this doc and that's what I want to clarify. So even if I define my function with def, will it work asynchronously? But such behavior is not good for this particular path operation function of mine. How do I make it work synchronously?

**Comment by ricky-sb** [comment]
>endpoint generates that many different OTPs  

Based on this, does that mean `check_for_non_expired_otp()` is returning `True` when you define the function with `async def`?

**Comment by assman** [comment]
> > endpoint generates that many different OTPs
> 
> Based on this, does that mean `check_for_non_expired_otp()` is returning `True` when you define the function with `async def`?

Sorry I didn't quite understand the question. Are you asking that if I declare the function with `async def`, the `check_for_non_expired_otp()` returns `True`?

The` check_for_non_expired_otp()` returns `True` or `False` depending on wether the database already has an OTP or not. It has nothing to do with wether I define the function as `def` or `async def`

**Comment by ricky-sb** [comment]
>I only wish to generate one OTP for a user at a time instead of multiple. 

It's generating multiple OTP codes when running asynchronously, right? But when you run it synchronously it generates a single code?  

In other words, if you `print(non_expired_otp)` and hammer the API with requests, do you see something like:

#### Async
* False
* False
* False 
* True
* False

#### Sync
* False
* True
* True 
* True

**Comment by assman** [comment]
I conducted the hammer tests and here are the results. For both the cases, I started with a blank database with no OTPs inside.

Result when using `def` - 

- False
- False
- False
- False
- False

Result when using `async def` - 

- False
- True
- True
- True
- True

**Comment by ricky-sb** [comment]
> I conducted the hammer tests and here are the results. For both the cases, I started with a blank database with no OTPs inside.
> 
> Result when using `def` -
> 
> * False
> * False
> * False
> * False
> * False
> 
> Result when using `async def` -
> 
> * False
> * True
> * True
> * True
> * True

Okay cool, you've got yourself a race condition. It depends on the database and the ORM you're using. But you'll want to wrap the `check_for_non_expired_otp()` and `generate_otp_for_user()` code in a single [`TRANSACTION`](https://www.postgresql.org/docs/8.3/tutorial-transactions.html).  

[More info here](https://techspot.zzzeek.org/2015/02/15/asynchronous-python-and-databases/), scroll down to "Database Code Handles Concurrency through ACID, Not In-Process Synchronization".

**Comment by assman** [comment]
> Okay cool, you've got yourself a race condition. It depends on the database and the ORM you're using. But you'll want to wrap the `check_for_non_expired_otp()` and `generate_otp_for_user()` code in a single [`TRANSACTION`](https://www.postgresql.org/docs/8.3/tutorial-transactions.html).
> 
> [More info here](https://techspot.zzzeek.org/2015/02/15/asynchronous-python-and-databases/), scroll down to "Database Code Handles Concurrency through ACID, Not In-Process Synchronization".

I see, thank you for your help. I am using MongoDB and pymongo as the driver.

Also, when I use `async def` instead of `def`, this unwanted behavior seems to go away. Why is that so?
